File: SymmetricTree101.java
import java.util.LinkedList;
import java.util.Queue;

//time complexity : O(n)
//space complexity :  O(n)
//leet code execution : successful
//steps : created helper function for traversal for left and right sub trees. recurrsion returns true only when left and right subtree vals match.
class SymmetricTree101 {

     public class TreeNode {
         int val;
         TreeNode left;
         TreeNode right;
         TreeNode(int x) { val = x; }}


    public boolean isSymmetric(TreeNode root){
        //if(root==null) return true;
        //return helper(root.left,root.right);

        if(root==null){
            return true;
        }

        Queue<TreeNode> queue =new LinkedList<>();
        queue.add(root.left);
        queue.add(root.right);

        while(!queue.isEmpty()){
            TreeNode nodeA,nodeB;
            nodeA = queue.poll();
            nodeB = queue.poll();

            if(nodeA==null && nodeB==null) continue;
            if(nodeA==null||nodeB==null)return false;
            if(nodeA.val!=nodeB.val) return false;
            queue.add(nodeA.left);
            queue.add(nodeB.right);
            queue.add(nodeA.right);
            queue.add(nodeB.left);
        }
        return true;
        }

/**
	private boolean helper(TreeNode left, TreeNode right) {
        if(left==null && right==null ){
            return true;    
        }
        if(left==null||right==null||left.val!=right.val) return false;

        return helper(left.left, right.right) && helper(left.right, right.left);

    }
    */

}

File: symmetric-tree_101.py
#Approach: checking left and right nodes and lefts node right value with right nodes left and viceversa
#Recursive
#Time Complexity O(N) for all N nodes
#Space Complexity O(H) height of tree
#It successfully runs on the leetcode


class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        return self.helper(root.left,root.right)
    
    def helper(self,left,right):
        if left and right:
            if left.val!=right.val: return  False
            return self.helper(left.left,right.right) and self.helper(left.right,right.left)
        else:
            return left==right


#Approach: checking lefts node right value with right nodes left and viceversa
#Iterative with stack
#Time Complexity O(N) for all N nodes
#Space Complexity O(H) height of tree stack in this case
#It successfully runs on the leetcode
class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        stack=[]
        if not root: return True
        stack.append((root.left,root.right))
     
        
        while stack:
            left, right = stack.pop()
            if not left and not right: continue
            if not left or not right or left.val != right.val:
                return False
            stack.append((left.left,right.right))
            stack.append((left.right,right.left))
            
        return True